### PR TITLE
V2

### LINKED ISSUE #5: Driver must be stateless or not a singleton
Current driver interface enforces driver to be stateful by `Initialize` method, which is a bit weird, as each registered driver is a singleton, so following code does not work as expected:
```go
fooDrv  := driver.New("mysql:/root:@(localhost:3306)/foo")(*mysql.Driver)
barDrv  := driver.New("mysql:/root:@(localhost:3306)/bar")(*mysql.Driver)

if fooDrv == barDrv {
       panic("surprise") // in other words, fooDrv will migrate bar database
}
```

### PR DESCRIPTION
**Pull request by josephbuchma**
Quite massive backward-incompatible refactoring :)
All drivers and journey cli are also updated.
This PR addresses all concerns discussed in #2 (in one way or another)

## Changes
- Removed all "async" stuff.
- Added type for encapsulating migrations functionality for greater flexibility and performance.
- Changed driver registration and initialization approach, fixed #5 
   - Removed `Initialize` method from Driver interface
   - Removed `FlenameExtension` from Driver interface
- Added driver.Locker interface, which can be optionally implemented to enable locking during migraitons.
   - Implemented for mysql driver
- Migrate now receives context.Context, and therefore can be cancelled.
- Added option to attach pre/post hooks for migrations.
- Added methods for applying / rolling back specific version.

### CODE REVIEW

**Comment by gravis** [comment]
A lot of things to review :)
Please give me a few days

**Comment by josephbuchma** [comment]
@gravis tests on master branch are failing
```
❯❯❯ make test                                                                   ⏎ master ✱
docker-compose run --rm go-test
Starting migrate_crate_1
=== RUN   TestCreate
--- PASS: TestCreate (0.24s)
        migrate_test.go:31: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:31: Test driver: mysql://root@tcp(172.17.0.9:3306)/migratetest
        migrate_test.go:31: Test driver: sqlite3:///tmp/migrate.db
=== RUN   TestReset
--- PASS: TestReset (0.73s)
        migrate_test.go:72: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:72: Test driver: mysql://root@tcp(172.17.0.9:3306)/migratetest
        migrate_test.go:72: Test driver: sqlite3:///tmp/migrate.db
=== RUN   TestDown
--- FAIL: TestDown (0.17s)
        migrate_test.go:110: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:140: Expected version 0, got 20171227092036
=== RUN   TestUp
--- FAIL: TestUp (0.05s)
        migrate_test.go:147: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:165: Expected version 0, got 20171227092036
=== RUN   TestRedo
--- PASS: TestRedo (1.03s)
        migrate_test.go:189: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:189: Test driver: mysql://root@tcp(172.17.0.9:3306)/migratetest
        migrate_test.go:189: Test driver: sqlite3:///tmp/migrate.db
=== RUN   TestMigrate
--- FAIL: TestMigrate (0.18s)
        migrate_test.go:229: Test driver: postgres://postgres@172.17.0.12:5432/template1?sslmode=disable
        migrate_test.go:267: Versions in db: [20171227092036 20171227092035 20171227092034 20171227092033 20171227092032 20171227092031 20171227092030 20171227092029 20171227092028 20171227092027 20171226235039 20171226235038 20060102150405]
        migrate_test.go:268: Expected version 0, got 20171227092036
FAIL
exit status 1
FAIL    github.com/db-journey/migrate   2.399s
?       github.com/db-journey/migrate/direction [no test files]
?       github.com/db-journey/migrate/driver    [no test files]
=== RUN   TestParseFilenameSchema
--- PASS: TestParseFilenameSchema (0.00s)
=== RUN   TestFiles
--- PASS: TestFiles (0.00s)
=== RUN   TestDuplicateFiles
--- PASS: TestDuplicateFiles (0.00s)
PASS
ok      github.com/db-journey/migrate/file      0.003s
?       github.com/db-journey/migrate/pipe      [no test files]
Makefile:7: recipe for target 'test' failed
make: *** [test] Error 1

```

**Comment by josephbuchma** [comment]
@gravis I added optional pre/post hooks for migraitons.
Useful for automating more complex migrations, and it also can be used for progress logging on client side (CLI).

**Comment by josephbuchma** [comment]
@gravis just want to note that I think I addressed all concerns discussed in #2 (in one way or another)

**Comment by josephbuchma** [comment]
Happy New Year !

Waiting for your feedback @gravis  :)

**Comment by josephbuchma** [comment]
ping

**Comment by gravis** [comment]
Hi there. I'm back to work today, sorry for the delay. 
I'll review your PR asap.
Thanks

File: migrate.go [hunk 1]
```diff
@@ -13,196 +13,182 @@ import (
...
+func WithHooks(pre, post func(f file.File) error) Option {
+	return func(h *Handle) error {
+		h.preHook = pre
+		h.postHook = post
+		return nil
 	}
+}
+
+// Handle encapsulates migrations functionality
+type Handle struct {
```
Comment (gravis) [inline migrate.go:31]: Shouldn't we name it `Handler`, instead of `Handle`?

**Review by gravis — APPROVED** [review]
LGTM.
`Handle` name is certainly not blocking.
Let's continue with changes in drivers, and we'll be good to merge

File: migrate.go (hunk 1 above)
Comment (josephbuchma) [reply migrate.go:31]: I don't like `Handler` b/c it's common name/pattern which usually refers to interface or function.
We can go conservative way and name it `Migrate` . I was simply looking for something that plays well with `migrate.` prefix and isn't counter-intuitive (at least), and `Handle` is the best what I came up with.

**Comment by josephbuchma** [comment]
round 2 @gravis  :smile: 
This time it's a bit more radical. I decided to fix #5 and it went a bit further.
All drivers are up to date in respective PRs

**Comment by josephbuchma** [comment]
@gravis now I think I'm happy with results. From my perspective it's ready for merge, and drivers can be polished with follow-up PRs.

**Comment by gravis** [comment]
Thanks for your tremendous work!

**Pull request merged**